mhs/mhs/common/state/work_description.py

```diff
@@ -0,0 +1,122 @@
+import json
+import utilities.integration_adaptors_logger as log
+
+
+logger = log.IntegrationAdaptorsLogger('STATE_MANAGER')
+
+
+class MessageStatus:
+    RECEIVED = 1
+    STARTED = 2
+    IN_OUTBOUND_WORKFLOW = 3
+
+
+DATA_KEY = 'MESSAGE_KEY'
+VERSION_KEY = 'VERSION'
+TIMESTAMP = 'TIMESTAMP'
+DATA = 'DATA'
+STATUS = 'STATUS'
+
+
+class OutOfDateVersionError(RuntimeError):
+    pass
+
+
+class EmptyWorkDescriptionError(RuntimeError):
+    pass
+
+
+class WorkDescription:
+
+    def __init__(self, persistence_store, table_name: str, store_data: dict):
+        if persistence_store is None:
+            raise ValueError('Expected persistence store')
+        if not table_name:
+            raise ValueError('Expected non empty table name')
+
+        self.persistence_store = persistence_store
+        self.table_name = table_name
+
+        data = store_data[DATA]
+        self.message_key = store_data[DATA_KEY]
+        self.version = data[VERSION_KEY]
+        self.timestamp = data[TIMESTAMP]
+        self.status = data[STATUS]
+
+    async def publish(self):
+        logger.info('011', 'Attempting to publish work description {key}', {'key': self.message_key})
+        logger.info('012', 'Retrieving latest work description to check version')
+        latest_data = await self.persistence_store.get(self.table_name, self.message_key)
```

Comment:
So a get is done first, then the version check is done, then based on that, the db is/isn't updated. If multiple things are doing this at the same time, then we've got a race condition. We just need to be careful that if we want to have a failsafe around multiple updates, then we're fine with this possible race condition, otherwise we should move the version check onto the database.

Reply:
Good point, I guess we'll have to remember to place the check on the database as well.

Reply:
I don't see this as having a check in 2 places. If we wanted to do it, I think we'd have to use [something like this](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html). Feels like a Dave question.